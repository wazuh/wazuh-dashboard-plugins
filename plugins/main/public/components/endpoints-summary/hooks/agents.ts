import { useState, useEffect } from 'react';
import { getTotalAgentsService } from '../services';

export const useGetTotalAgents = () => {
  const [totalAgents, setTotalAgents] = useState<number>();
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getTotalAgents = async () => {
    try {
      setIsLoading(true);
      const totalAgents = await getTotalAgentsService();
      setTotalAgents(totalAgents);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getTotalAgents();
  }, []);

  return {
    isLoading,
    totalAgents,
    error,
  };
};
