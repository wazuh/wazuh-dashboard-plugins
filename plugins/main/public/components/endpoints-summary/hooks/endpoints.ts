import { useState, useEffect } from 'react';
import { WzRequest } from '../../../react-services/wz-request';

export const useGetTotalEndpoints = () => {
  const [totalEndpoints, setTotalEndpoints] = useState<number>();
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getTotalEndpoints = async () => {
    try {
      setIsLoading(true);
      const {
        data: {
          data: { total_affected_items },
        },
      } = await WzRequest.apiReq('GET', '/agents', {
        params: { limit: 1, q: 'id!=000' },
      });
      setTotalEndpoints(total_affected_items);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getTotalEndpoints();
  }, []);

  return {
    isLoading,
    totalEndpoints,
    error,
  };
};
