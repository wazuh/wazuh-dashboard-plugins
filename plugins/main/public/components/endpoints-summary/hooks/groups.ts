import { useState, useEffect } from 'react';
import { getGroupsService } from '../services';

export const useGetGroups = () => {
  const [groups, setGroups] = useState<string[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const getGroups = async () => {
    try {
      setIsLoading(true);
      const { affected_items } = await getGroupsService();
      const groups = affected_items.map(item => item.name);
      setGroups(groups);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    getGroups();
  }, []);

  return {
    isLoading,
    groups,
    error,
  };
};
