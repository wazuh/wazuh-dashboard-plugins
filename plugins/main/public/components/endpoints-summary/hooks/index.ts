export { useGetTotalAgents } from './agents';
export { useGetGroups } from './groups';
