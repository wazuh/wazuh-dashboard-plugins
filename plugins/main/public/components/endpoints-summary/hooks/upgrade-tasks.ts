import { useState, useEffect } from 'react';
import { getTasks } from '../services';
import { API_NAME_TASK_STATUS } from '../../../../common/constants';

const beforeMinutes = 60;

export const useGetUpgradeTasks = (reload: any) => {
  const [totalInProgressTasks, setTotalInProgressTasks] = useState<number>(0);
  const [getInProgressIsLoading, setGetInProgressIsLoading] = useState(true);
  const [getInProgressError, setGetInProgressError] = useState();

  const [totalSuccessTasks, setTotalSuccessTasks] = useState<number>(0);
  const [getSuccessIsLoading, setSuccessIsLoading] = useState(true);
  const [getSuccessError, setGetSuccessError] = useState();

  const [totalErrorUpgradeTasks, setTotalErrorUpgradeTasks] =
    useState<number>(0);
  const [getErrorIsLoading, setErrorIsLoading] = useState(true);
  const [getErrorTasksError, setGetErrorTasksError] = useState();

  const [totalTimeoutUpgradeTasks, setTotalTimeoutUpgradeTasks] =
    useState<number>(0);
  const [getTimeoutIsLoading, setTimeoutIsLoading] = useState(true);
  const [getTimeoutError, setGetTimeoutError] = useState();

  const datetime = new Date();
  datetime.setMinutes(datetime.getMinutes() - beforeMinutes);
  const formattedDate = datetime.toISOString();
  const timeFilter = `last_update_time>${formattedDate}`;

  const getUpgradeStatus = async (
    status: string,
    setIsLoading: (isLoading: boolean) => void,
    setTotalTasks: (totalTasks: number) => void,
    setError: (error) => void,
    q?: string,
  ) => {
    try {
      setIsLoading(true);
      const { total_affected_items } = await getTasks({
        status,
        command: 'upgrade',
        limit: 1,
        q,
      });
      setTotalTasks(total_affected_items);
      setError(undefined);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  const getUpgradesInProgress = async () =>
    await getUpgradeStatus(
      API_NAME_TASK_STATUS.IN_PROGRESS,
      setGetInProgressIsLoading,
      setTotalInProgressTasks,
      setGetInProgressError,
    );

  const getUpgradesSuccess = async () => {
    await getUpgradeStatus(
      API_NAME_TASK_STATUS.DONE,
      setSuccessIsLoading,
      setTotalSuccessTasks,
      setGetSuccessError,
      timeFilter,
    );
  };

  const getUpgradesError = async () => {
    await getUpgradeStatus(
      API_NAME_TASK_STATUS.FAILED,
      setErrorIsLoading,
      setTotalErrorUpgradeTasks,
      setGetErrorTasksError,
      timeFilter,
    );
  };

  const getUpgradeTimeout = async () => {
    await getUpgradeStatus(
      API_NAME_TASK_STATUS.TIMEOUT,
      setTimeoutIsLoading,
      setTotalTimeoutUpgradeTasks,
      setGetTimeoutError,
      timeFilter,
    );
  };

  const fetchData = async () => {
    await getUpgradesInProgress();
    await getUpgradesSuccess();
    await getUpgradesError();
    await getUpgradeTimeout();
  };

  useEffect(() => {
    fetchData();

    const intervalId = setInterval(getUpgradesInProgress, 3000);

    if (totalInProgressTasks === 0) {
      clearInterval(intervalId);
    }

    return () => clearInterval(intervalId);
  }, [totalInProgressTasks, reload]);

  return {
    getInProgressIsLoading,
    totalInProgressTasks,
    getInProgressError,
    getSuccessIsLoading,
    totalSuccessTasks,
    getSuccessError,
    getErrorIsLoading,
    totalErrorUpgradeTasks,
    getErrorTasksError,
    getTimeoutIsLoading,
    totalTimeoutUpgradeTasks,
    getTimeoutError,
  };
};
