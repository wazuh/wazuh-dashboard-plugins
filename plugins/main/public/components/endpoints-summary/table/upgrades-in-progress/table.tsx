import React from 'react';
import { EuiHealth } from '@elastic/eui';
import { TableWzAPI } from '../../../common/tables';
import { formatUIDate } from '../../../../react-services/time-service';
import {
  API_NAME_TASK_STATUS,
  SEARCH_BAR_WQL_VALUE_SUGGESTIONS_COUNT,
  UI_TASK_STATUS,
} from '../../../../../common/constants';
import { WzRequest } from '../../../../react-services/wz-request';
import { get as getLodash } from 'lodash';

export const AgentUpgradesTable = () => {
  const datetime = new Date();
  datetime.setMinutes(datetime.getMinutes() - 60);
  const formattedDate = datetime.toISOString();

  const defaultFilters = {
    q: `last_update_time>${formattedDate}`,
  };

  const searchBarWQLOptions = {
    implicitQuery: {
      query: 'id!=000',
      conjunction: ';',
    },
  };

  return (
    <TableWzAPI
      title='Upgrade agent tasks'
      endpoint='/tasks/status'
      tableColumns={[
        {
          field: 'task_id',
          name: 'Task id',
          sortable: true,
          searchable: true,
          show: true,
        },
        {
          field: 'agent_id',
          name: 'Agent id',
          sortable: true,
          searchable: true,
          show: true,
        },
        {
          field: 'create_time',
          name: 'Create',
          sortable: true,
          searchable: true,
          show: true,
          render: value => formatUIDate(value),
        },
        {
          field: 'last_update_time',
          name: 'Last update',
          sortable: true,
          searchable: true,
          show: true,
          render: value => formatUIDate(value),
        },
        {
          field: 'status',
          name: 'Status',
          width: '100px',
          sortable: true,
          searchable: true,
          show: true,
          render: value => (
            <EuiHealth
              color={
                value === API_NAME_TASK_STATUS.DONE
                  ? 'success'
                  : value === API_NAME_TASK_STATUS.IN_PROGRESS
                  ? 'warning'
                  : 'danger'
              }
            >
              {value}
            </EuiHealth>
          ),
        },
        {
          field: 'error_message',
          name: 'Error',
          show: true,
        },
      ]}
      tableInitialSortingField='last_update_time'
      tableInitialSortingDirection='desc'
      tablePageSizeOptions={[10, 25, 50, 100]}
      filters={defaultFilters}
      searchTable
      searchBarWQL={{
        suggestions: {
          field(currentValue) {
            return [
              { label: 'agent_id', description: 'filter by agent id' },
              { label: 'status', description: 'filter by status' },
              {
                label: 'create_time',
                description: 'filter by creation date',
              },
              {
                label: 'last_update_time',
                description: 'filter by last update date',
              },
              { label: 'task_id', description: 'filter by task id' },
            ];
          },
          value: async (currentValue, { field }) => {
            try {
              switch (field) {
                case 'status':
                  return UI_TASK_STATUS.map(status => ({
                    label: status,
                  }));
                case 'agent_id': {
                  const response = await WzRequest.apiReq('GET', '/agents', {
                    params: {
                      distinct: true,
                      limit: SEARCH_BAR_WQL_VALUE_SUGGESTIONS_COUNT,
                      select: 'id',
                      sort: `+id`,
                      ...(currentValue
                        ? {
                            q: `${searchBarWQLOptions.implicitQuery.query}${searchBarWQLOptions.implicitQuery.conjunction}id~${currentValue}`,
                          }
                        : {
                            q: `${searchBarWQLOptions.implicitQuery.query}`,
                          }),
                    },
                  });
                  return response?.data?.data.affected_items.map(item => ({
                    label: getLodash(item, 'id'),
                  }));
                }
              }
            } catch (error) {
              return [];
            }
          },
        },
        validate: {
          value: ({ formattedValue, value: rawValue }, { field }) => {
            const value = formattedValue ?? rawValue;
            if (value) {
              if (['create_time', 'last_update_time'].includes(field)) {
                const isCorrectDate =
                  /^\d{4}-\d{2}-\d{2}([ T]\d{2}:\d{2}:\d{2}(.\d{1,6})?Z?)?$/.test(
                    value,
                  );
                return isCorrectDate
                  ? undefined
                  : `"${value}" is not a expected format. Valid formats: YYYY-MM-DD, YYYY-MM-DD HH:mm:ss, YYYY-MM-DDTHH:mm:ss, YYYY-MM-DDTHH:mm:ssZ.`;
              }
            }
          },
        },
      }}
      tableProps={{
        tableLayout: 'auto',
      }}
    />
  );
};
