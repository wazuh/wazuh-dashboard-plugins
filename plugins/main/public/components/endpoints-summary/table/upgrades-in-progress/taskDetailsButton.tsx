import React, { useState, memo } from 'react';
import { EuiHealth, EuiIconTip } from '@elastic/eui';
import { TableWzAPI } from '../../../common/tables';
import { formatUIDate } from '../../../../react-services/time-service';
import {
  API_NAME_TASK_STATUS,
  SEARCH_BAR_WQL_VALUE_SUGGESTIONS_COUNT,
  UI_TASK_STATUS,
} from '../../../../../common/constants';
import { WzRequest } from '../../../../react-services/wz-request';
import { get as getLodash, uniqBy as uniqByLodash } from 'lodash';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalBody,
  EuiModalFooter,
  EuiButton,
  EuiButtonEmpty,
} from '@elastic/eui';

export const AgentUpgradesTaskDetailsButton = memo(() => {
  const [isModalVisible, setIsModalVisible] = useState(false);

  const datetime = new Date();
  datetime.setMinutes(datetime.getMinutes() - 60);
  const formattedDate = datetime.toISOString();

  const defaultFilters = {
    q: `last_update_time>${formattedDate}`,
  };

  const handleOnCloseModal = () => setIsModalVisible(false);

  return (
    <>
      <EuiButtonEmpty
        color='primary'
        onClick={() => setIsModalVisible(true)}
        iconType='eye'
      >
        Task details
      </EuiButtonEmpty>
      {isModalVisible ? (
        <EuiModal onClose={handleOnCloseModal} maxWidth={false}>
          <EuiModalHeader />
          <EuiModalBody>
            <TableWzAPI
              title='Upgrade agent tasks'
              endpoint='/tasks/status'
              tableColumns={[
                {
                  field: 'task_id',
                  name: 'Task id',
                  sortable: true,
                  searchable: true,
                  show: true,
                },
                {
                  field: 'agent_id',
                  name: 'Agent id',
                  sortable: true,
                  searchable: true,
                  show: true,
                },
                {
                  field: 'create_time',
                  name: (
                    <span>
                      Create{' '}
                      <EuiIconTip
                        content='This is not searchable through a search term.'
                        size='s'
                        color='subdued'
                        type='alert'
                      />
                    </span>
                  ),
                  sortable: true,
                  searchable: false,
                  show: true,
                  render: value => formatUIDate(value),
                },
                {
                  field: 'last_update_time',
                  name: (
                    <span>
                      Last update{' '}
                      <EuiIconTip
                        content='This is not searchable through a search term.'
                        size='s'
                        color='subdued'
                        type='alert'
                      />
                    </span>
                  ),
                  sortable: true,
                  searchable: false,
                  show: true,
                  render: value => formatUIDate(value),
                },
                {
                  field: 'status',
                  name: 'Status',
                  width: '100px',
                  sortable: true,
                  searchable: true,
                  show: true,
                  render: value => (
                    <EuiHealth
                      color={
                        value === API_NAME_TASK_STATUS.DONE
                          ? 'success'
                          : value === API_NAME_TASK_STATUS.IN_PROGRESS
                          ? 'warning'
                          : 'danger'
                      }
                    >
                      {value}
                    </EuiHealth>
                  ),
                },
                {
                  field: 'error_message',
                  name: 'Error',
                  show: true,
                  searchable: true,
                },
              ]}
              tableInitialSortingField='last_update_time'
              tableInitialSortingDirection='desc'
              tablePageSizeOptions={[10, 25, 50, 100]}
              filters={defaultFilters}
              downloadCsv
              showReload
              showFieldSelector
              searchTable
              searchBarWQL={{
                suggestions: {
                  field(currentValue) {
                    return [
                      {
                        label: 'agent_id',
                        description: 'filter by agent id',
                      },
                      { label: 'status', description: 'filter by status' },
                      {
                        label: 'create_time',
                        description: 'filter by creation date',
                      },
                      {
                        label: 'last_update_time',
                        description: 'filter by last update date',
                      },
                      { label: 'task_id', description: 'filter by task id' },
                    ];
                  },
                  value: async (currentValue, { field }) => {
                    try {
                      switch (field) {
                        case 'status':
                          return UI_TASK_STATUS.map(status => ({
                            label: status,
                          }));
                        default: {
                          const response = await WzRequest.apiReq(
                            'GET',
                            '/tasks/status',
                            {
                              params: {
                                limit: SEARCH_BAR_WQL_VALUE_SUGGESTIONS_COUNT,
                                select: field,
                                sort: `+${field}`,
                                ...(currentValue
                                  ? {
                                      q: `${field}~${currentValue}`,
                                    }
                                  : {}),
                              },
                            },
                          );
                          const suggestionValues =
                            response?.data?.data.affected_items.map(item => ({
                              label: getLodash(item, field),
                            }));
                          return uniqByLodash(suggestionValues, 'label');
                        }
                      }
                    } catch (error) {
                      return [];
                    }
                  },
                },
                validate: {
                  value: ({ formattedValue, value: rawValue }, { field }) => {
                    const value = formattedValue ?? rawValue;
                    if (value) {
                      if (['create_time', 'last_update_time'].includes(field)) {
                        const isCorrectDate =
                          /^\d{4}-\d{2}-\d{2}([ T]\d{2}:\d{2}:\d{2}(.\d{1,6})?Z?)?$/.test(
                            value,
                          );
                        return isCorrectDate
                          ? undefined
                          : `"${value}" is not a expected format. Valid formats: YYYY-MM-DD, YYYY-MM-DD HH:mm:ss, YYYY-MM-DDTHH:mm:ss, YYYY-MM-DDTHH:mm:ssZ.`;
                      }
                    }
                  },
                },
              }}
              tableProps={{
                tableLayout: 'auto',
              }}
            />
          </EuiModalBody>
          <EuiModalFooter>
            <EuiButton onClick={handleOnCloseModal} fill>
              Close
            </EuiButton>
          </EuiModalFooter>
        </EuiModal>
      ) : null}
    </>
  );
});
