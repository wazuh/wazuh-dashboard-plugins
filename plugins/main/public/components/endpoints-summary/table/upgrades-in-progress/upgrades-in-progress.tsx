import React, { useState, useEffect } from 'react';
import {
  EuiPanel,
  EuiProgress,
  EuiText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIconTip,
  EuiSpacer,
  EuiButtonEmpty,
  EuiButtonIcon,
} from '@elastic/eui';
import { useGetUpgradeTasks } from '../../hooks';
import {
  API_NAME_TASK_STATUS,
  UI_LOGGER_LEVELS,
} from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';

interface AgentUpgradesInProgress {
  reload: any;
  setIsModalVisible: (isModalVisible: boolean) => void;
  isPanelClosed: boolean;
  setIsPanelClosed: (isPanelClosed: boolean) => void;
  allowGetTasks: boolean;
}

export const AgentUpgradesInProgress = ({
  reload,
  setIsModalVisible,
  isPanelClosed,
  setIsPanelClosed,
  allowGetTasks,
}: AgentUpgradesInProgress) => {
  const [isUpgrading, setIsUpgrading] = useState(false);

  const {
    totalInProgressTasks = 0,
    getInProgressError = undefined,
    totalSuccessTasks = 0,
    getSuccessError = undefined,
    totalErrorUpgradeTasks = 0,
    getErrorTasksError = undefined,
    totalTimeoutUpgradeTasks = 0,
    getTimeoutError = undefined,
  } = allowGetTasks ? useGetUpgradeTasks(reload) : {};

  useEffect(() => {
    if (totalInProgressTasks > 0) {
      setIsUpgrading(true);
    }
  }, [totalInProgressTasks]);

  const showErrorToast = (status: string, error: any) => {
    API_NAME_TASK_STATUS;
    const options = {
      context: `AgentUpgradesInProgress.useGetUpgradeTasks`,
      level: UI_LOGGER_LEVELS.ERROR,
      severity: UI_ERROR_SEVERITIES.BUSINESS,
      store: true,
      error: {
        error,
        message: error.message || error,
        title: `Could not get upgrade tasks: ${status}`,
      },
    };
    getErrorOrchestrator().handleError(options);
  };

  if (getInProgressError) {
    showErrorToast(API_NAME_TASK_STATUS.IN_PROGRESS, getInProgressError);
  }

  if (getSuccessError) {
    showErrorToast(API_NAME_TASK_STATUS.DONE, getSuccessError);
  }

  if (getErrorTasksError) {
    showErrorToast(API_NAME_TASK_STATUS.FAILED, getErrorTasksError);
  }

  if (getTimeoutError) {
    showErrorToast(API_NAME_TASK_STATUS.TIMEOUT, getTimeoutError);
  }

  const showTasks = isUpgrading || totalSuccessTasks || totalErrorUpgradeTasks;
  if (isPanelClosed || !showTasks) return null;

  return (
    <EuiPanel color='subdued'>
      <EuiFlexGroup
        gutterSize='s'
        alignItems='flexStart'
        wrap={false}
        responsive={false}
      >
        <EuiFlexGroup
          gutterSize='s'
          alignItems='center'
          wrap={false}
          responsive={false}
        >
          <EuiFlexItem grow={false}>
            <EuiText>Upgrade tasks</EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButtonEmpty
              color='primary'
              onClick={() => setIsModalVisible(true)}
              iconType='eye'
            >
              Task details
            </EuiButtonEmpty>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiFlexItem grow={false}>
          <EuiButtonIcon
            onClick={() => setIsPanelClosed(true)}
            color='text'
            iconType='cross'
            aria-label='Close'
          />
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer size='s' />
      <EuiFlexGroup gutterSize='s' alignItems='center'>
        {totalInProgressTasks > 0 ? (
          <EuiFlexItem grow={false}>
            <EuiPanel paddingSize='s' style={{ position: 'relative' }}>
              <EuiProgress size='xs' color='warning' position='absolute' />
              <EuiText size='s'>
                <b>{totalInProgressTasks}</b>
                {` ${API_NAME_TASK_STATUS.IN_PROGRESS}`}
              </EuiText>
            </EuiPanel>
          </EuiFlexItem>
        ) : null}
        {totalSuccessTasks > 0 ? (
          <EuiFlexItem grow={false}>
            <EuiPanel paddingSize='s' style={{ position: 'relative' }}>
              <EuiProgress
                value={100}
                max={100}
                size='xs'
                color='success'
                position='absolute'
              />
              <span>
                <EuiText size='s'>
                  <b>{totalSuccessTasks}</b>
                  {` ${API_NAME_TASK_STATUS.DONE} `}
                  <EuiIconTip content='Last 60 minutes' color='primary' />
                </EuiText>
              </span>
            </EuiPanel>
          </EuiFlexItem>
        ) : null}
        {totalErrorUpgradeTasks > 0 ? (
          <EuiFlexItem grow={false}>
            <EuiPanel paddingSize='s' style={{ position: 'relative' }}>
              <EuiProgress
                value={100}
                max={100}
                size='xs'
                color='danger'
                position='absolute'
              />
              <span>
                <EuiText size='s'>
                  <b>{totalErrorUpgradeTasks}</b>
                  {` ${API_NAME_TASK_STATUS.FAILED} `}
                  <EuiIconTip content='Last 60 minutes' color='primary' />
                </EuiText>
              </span>
            </EuiPanel>
          </EuiFlexItem>
        ) : null}
        {totalTimeoutUpgradeTasks > 0 ? (
          <EuiFlexItem grow={false}>
            <EuiPanel paddingSize='s' style={{ position: 'relative' }}>
              <EuiProgress
                value={100}
                max={100}
                size='xs'
                color='subdued'
                position='absolute'
              />
              <span>
                <EuiText size='s'>
                  <b>{totalTimeoutUpgradeTasks}</b>
                  {` ${API_NAME_TASK_STATUS.TIMEOUT} `}
                  <EuiIconTip content='Last 60 minutes' color='primary' />
                </EuiText>
              </span>
            </EuiPanel>
          </EuiFlexItem>
        ) : null}
      </EuiFlexGroup>
    </EuiPanel>
  );
};
