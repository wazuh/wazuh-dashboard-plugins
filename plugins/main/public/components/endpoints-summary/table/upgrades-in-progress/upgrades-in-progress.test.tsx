import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { AgentUpgradesInProgress } from './upgrades-in-progress';
import { useGetUpgradeTasks } from '../../hooks';

jest.mock('../../hooks', () => ({
  useGetUpgradeTasks: jest.fn(),
}));

jest.mock('../../../../react-services/common-services', () => ({
  getErrorOrchestrator: () => ({
    handleError: () => {},
  }),
}));

describe('AgentUpgradesInProgress component', () => {
  test('should return the component', async () => {
    (useGetUpgradeTasks as jest.Mock).mockReturnValue({
      getInProgressIsLoading: false,
      totalInProgressTasks: 5,
      getErrorIsLoading: false,
      totalErrorUpgradeTasks: 2,
    });

    const { container, getByText } = render(
      <AgentUpgradesInProgress
        reload={0}
        setIsModalVisible={() => {}}
        isPanelClosed={false}
        setIsPanelClosed={() => {}}
        allowGetTasks={true}
      />,
    );

    expect(container).toMatchSnapshot();

    const inProgressValue = getByText('5');
    expect(inProgressValue).toBeInTheDocument();
    const inProgressText = getByText('In progress');
    expect(inProgressText).toBeInTheDocument();

    const failedValue = getByText('2');
    expect(failedValue).toBeInTheDocument();
    const failedText = getByText('Failed');
    expect(failedText).toBeInTheDocument();
  });

  test('should show upgrade tasks modal', async () => {
    const { getByRole } = render(
      <AgentUpgradesInProgress
        reload={0}
        setIsModalVisible={() => {}}
        isPanelClosed={false}
        setIsPanelClosed={() => {}}
        allowGetTasks={true}
      />,
    );

    const openModalButton = getByRole('button', { name: 'Task details' });
    expect(openModalButton).toBeInTheDocument();

    act(() => {
      fireEvent.click(openModalButton);
    });
  });
});
