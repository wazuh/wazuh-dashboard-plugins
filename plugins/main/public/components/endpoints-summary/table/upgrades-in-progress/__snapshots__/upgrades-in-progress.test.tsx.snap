// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentUpgradesInProgress component should return the component 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--subdued euiPanel--noShadow euiPanel--noBorder"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiText euiText--medium"
          >
            Upgrade tasks
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--primary"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonEmpty__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiButtonEmpty__text"
              >
                Task details
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          aria-label="Close"
          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiPanel euiPanel--paddingSmall euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          style="position: relative;"
        >
          <div
            class="euiProgress euiProgress--indeterminate euiProgress--xs euiProgress--warning euiProgress--absolute"
          />
          <div
            class="euiText euiText--small"
          >
            <b>
              5
            </b>
             In progress
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiPanel euiPanel--paddingSmall euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          style="position: relative;"
        >
          <progress
            aria-hidden="false"
            class="euiProgress euiProgress--native euiProgress--xs euiProgress--danger euiProgress--absolute"
            max="100"
            value="100"
          />
          <span>
            <div
              class="euiText euiText--small"
            >
              <b>
                2
              </b>
               Failed 
              <span
                class="euiToolTipAnchor"
              >
                <svg
                  aria-hidden="true"
                  aria-label="Info"
                  class="euiIcon euiIcon--medium euiIcon--primary euiIcon-isLoading"
                  focusable="true"
                  height="16"
                  role="img"
                  tabindex="0"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                  />
                </svg>
              </span>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
